lib/Conversion/HWArithToHW/HWArithToHW.cpp

```diff
@@ -253,23 +255,179 @@ struct BinaryOpLowering : public OpConversionPattern<BinOp> {
+            })
+            .Case<hw::StructType>([this](auto type) {
+              return llvm::any_of(type.getElements(), [this](auto element) {
+                return this->hasSignednessSemantics(element.type);
+              });
+            })
+            .Default([](auto type) { return false; });
+
+    if (match) {
+      // Prime the conversion cache by pre-converting the type.
+      conversionCache[type] = {removeSignedness(type), true};
+    } else {
+      // Prime the conversion cache by not converting the type - this prevents
+      // iterating through the type on future removeSignedness calls for this
+      // type.
+      conversionCache[type] = {type, false};
+    }
+
+    return match;
+  }
+
+  bool hasSignednessSemantics(TypeRange types) {
+    return llvm::any_of(types,
+                        [this](Type t) { return hasSignednessSemantics(t); });
+  }
+
+private:
+  // Memoizations for signedness info and conversions.
+  struct ConvertedType {
+    Type type;
+    bool hadSignednessSemantics;
+  };
+  llvm::DenseMap<Type, ConvertedType> conversionCache;
+};
+
+HWArithToHWTypeConverter::HWArithToHWTypeConverter() {
+  // Pass any type through the signedness remover.
+  addConversion([this](Type type) { return removeSignedness(type); });
+}
+
+// Adds the ArgResOpConversion for 'TOp' to the set of conversion patterns, as
+// well as a legality check on the conversion status of the op's operands and
+// results.
+template <typename TOp>
+static void addOperandConversion(ConversionTarget &target,
+                                 RewritePatternSet &patterns,
+                                 HWArithToHWTypeConverter &typeConverter) {
+  patterns.add<ArgResOpConversion<TOp>>(typeConverter, patterns.getContext());
+  target.addDynamicallyLegalOp<TOp>([&](TOp op) {
+    return !typeConverter.hasSignednessSemantics(op->getOperandTypes()) &&
+           !typeConverter.hasSignednessSemantics(op->getResultTypes());
+  });
+}
+
+template <typename TOp, typename TOp2, typename... OpTs>
+static void addOperandConversion(ConversionTarget &target,
+                                 RewritePatternSet &patterns,
+                                 HWArithToHWTypeConverter &typeConverter) {
+  addOperandConversion<TOp>(target, patterns, typeConverter);
+  addOperandConversion<TOp2, OpTs...>(target, patterns, typeConverter);
+}
+
 } // namespace
 
 //===----------------------------------------------------------------------===//
 // Pass driver
 //===----------------------------------------------------------------------===//
 
 namespace {
+
 class HWArithToHWPass : public HWArithToHWBase<HWArithToHWPass> {
 public:
   void runOnOperation() override {
     ModuleOp module = getOperation();
 
     ConversionTarget target(getContext());
     RewritePatternSet patterns(&getContext());
+    HWArithToHWTypeConverter typeConverter;
+    mlir::populateFunctionOpInterfaceTypeConversionPattern<hw::HWModuleOp>(
+        patterns, typeConverter);
```

Comment:
If that's the case, then it might be worth considering introducing analogously to `addOperandConversion` a templated `addSignatureConversion` to apply both `populateFunctionOpInterfaceTypeConversionPattern` as well as `target.addDynamicallyLegalOp`.

Reply:
factored out into `addSignatureConversion`.